typing/typecore.ml

```diff
@@ -210,60 +210,79 @@ type tail_or_nontail = Tail | Nontail
     escaping_context : escaping_context option;
-    mode : Value_mode.t }
+    mode : Value_mode.t;
+    tuple_modes : Value_mode.t list; }
 
-let mode_tail mode =
+let mode_return mode =
   let position = Tail in
   let escaping_context = Some Return in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_var () =
   let position = Nontail in
   let escaping_context = None in
   let mode = Value_mode.newvar () in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_local =
   let position = Nontail in
   let escaping_context = None in
   let mode = Value_mode.local in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_global =
   let position = Nontail in
   let escaping_context = None in
   let mode = Value_mode.global in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_subcomponent expected_mode =
-  let mode = Value_mode.regional_to_global expected_mode.mode in
   let position = Nontail in
   let escaping_context = None in
-  { position; escaping_context; mode }
+  let mode = Value_mode.regional_to_global expected_mode.mode in
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_tailcall_function mode =
   let position = Nontail in
   let escaping_context = Some Tailcall_function in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_tailcall_argument mode =
   let position = Nontail in
   let escaping_context = Some Tailcall_argument in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
 
 let mode_partial_application expected_mode =
-  let mode = Value_mode.regional_to_global expected_mode.mode in
   let position = Nontail in
   let escaping_context = Some Partial_application in
-  { position; escaping_context; mode }
+  let mode = Value_mode.regional_to_global expected_mode.mode in
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
+
+let mode_trywith expected_mode =
+  { expected_mode with position = Nontail }
 
 let mode_nontail mode =
   let position = Nontail in
   let escaping_context = None in
-  { position; escaping_context; mode }
+  let tuple_modes = [] in
+  { position; escaping_context; mode; tuple_modes }
+
+let mode_tuple mode tuple_modes =
+  let position = Nontail in
+  let escaping_context = None in
+  { position; escaping_context; mode; tuple_modes }
 
 let submode ~loc ~env mode expected_mode =
-  match Types.Value_mode.submode mode expected_mode.mode with
+  let modes = expected_mode.mode :: expected_mode.tuple_modes in
+  match Types.Value_mode.submode_meet mode modes  with
```

Comment:
Due to the invariant mentioned above, this does a redundant check when there are tuple modes. Consider instead:
```ocaml
match expected_mode.tuple_modes with
| [] -> Types.Value_mode.submode mode expected_mode.mode
| tuples -> Types.Value_mode.submode_meet mode tuples
```